README.md
# Python-Carputer project

## Carputer project

- **This repository contains Python code used in the Carputer project.

https://github.com/billdoerr/Hardware-Carputer
https://github.com/billdoerr/Python-Carputer
https://github.com/billdoerr/Android-Carputer  


CHANGELOG.md
# Changelog
All notable changes to this Carputer Python code project will be documented in this file.
File:  D:\Users\bdoerr\Development\Python\PycharmProjects\carputer\CHANGELOG.md


## [Unreleased]
### v1.x (NOT STARTED)
#### Added
#### Changed
#### Removed



## [Released]
### v1.2.1 (9Jan2022)
#### Added
#### Changed  
- [x] New timesync.sh. Resolves issue #6 (Camera-Rear node not syncing with master). SSH certificate not working for some reason. Another script modification is to use the 'sshpass' package and just hard code the password. Doesn't matter, RPi's have no security. 
- [x] sudo apt-get install sshpass
- [x] New timesync.sh
- [x] Disable systemd-timesyncd
- [x] sudo chmod 777 timesync.sh
- [x] Update firmware version from v1.2 to v1.2.1.
#### Removed


## [Released]
### v1.2 (18Apr2021)
#### Added
- [x] juice4halt.py. This is a modified version of listen_for_shutdown.py where the code for the button push is removed.
#### Changed
#### Removed


## [Released]
### v1.1 (4Jul2019)
#### Added
#### Changed
- [x] Change version tag in listen_for_shutdown.py:  __version__ = '1.1'
- [x] listen_for_shutdown.sh:  Removed this section of code in start() function.
        # Remove log file upon start.  Used only for debugging.
        # sudo rm -f /var/log/carputer/listen_for_shutdown.log      # Addressed in python script filemode='w'
- [x] listen_for_shutdown.py:  Added "filemode='w'" in log_init(). file_handler = logging.FileHandler('/var/log/carputer/listen_for_shutdown.log', 'w')  Opens a file for writing only. Overwrites the file if the file exists. If the file does not exist, creates a new file for writing.    
- [x] listen_for_shutdown.py:  Changed log record formatting. 
- [x] listen_for_shutdown.py:  filename = '/var/lib/misc/dnsmasq.leases' does not contain ip's of nodes as expected.  Created filename = '/etc/carputer/nodes.config' which contains
                               single node '192.168.4.5' as temporary work around.  Python script changed to read the 'nodes' file instead of 'dnsmasq.leases'.
#### Removed


## [Released]
### v1.0 (17May2019)
#### Added
- [x]  listen_for_shutdown.py - Python script that listens for button press and after short period, ~5 seconds, will issue a shutdown command to slave and master nodes.
        Bash shell script also included so that this starts upon boot-up.

		• sudo apt install python3-paramiko     # install required ssh package
        
		• cd /etc/init.d
		• sudo vi listen_for_shutdown.sh
		• sudo chmod +x listen_for_shutdown.sh
		• sudo mkdir /var/log/carputer
		• cd /usr/local/bin
		• sudo vi listen_for_shutdown.py
		• sudo chmod +x listen_for_shutdown.py
		• sudo update-rc.d listen_for_shutdown.sh defaults
		• sudo reboot
		• cd /var/log/carputer/
		• ls
        • tail -f listen_for_shutdown.log
        
- [x] heartbeat.py - Python script that turns on/off led once per second to indicate that system is running.
        Bash shell script also included so that this starts upon boot-up.

		• cd /etc/init.d
		• sudo vi heartbeat.sh
		• sudo chmod +x heartbeat.sh
		• sudo mkdir /var/log/carputer
		• cd /usr/local/bin
		• sudo vi heartbeat.py
		• sudo chmod +x heartbeat.py
		• sudo update-rc.d heartbeat.sh defaults
		• sudo reboot
		• cd /var/log/carputer/
		• ls
		• tail -f heartbeat.log

#### Changed
#### Removed
		



heartbeat.py
# !/usr/bin/python3

"""
    heartbeat.py

    Blinks LED on/off indicating that the RPi is still running.

"""

__version__ = '1.0'
__author__ = 'Bill Doerr'


import RPi.GPIO as GPIO
from time import sleep


# Define some constants
CHANNEL = 13        # Pinout #13, GPIO #27
HEARTBEAT = 0.5     # Value in seconds


# GPIO initialization
def gpio_init():
    # Using RPi pin out.  GPIO Pin #37.  Button connected to GPIO -> Pin #37 and GND -> Pin #38
    GPIO.setmode(GPIO.BOARD)
    GPIO.setwarnings(False)
    GPIO.setup(CHANNEL, GPIO.OUT)


def main():

    gpio_init()

    try:
        # Loop while computer's power is on
        while True:
            GPIO.output(CHANNEL, GPIO.HIGH)
            sleep(HEARTBEAT)
            GPIO.output(CHANNEL, GPIO.LOW)
            sleep(HEARTBEAT)

    except KeyboardInterrupt:
        pass

    except:
        # This catches ALL other exceptions including errors.
        # You won't get any error messages for debugging
        # so only use it once your code is working.
        pass

    finally:
        GPIO.output(CHANNEL, GPIO.LOW)
        GPIO.cleanup()  # This ensures a clean exit.  Enables all pins for INPUT, but only for this program.


if __name__ == "__main__":
    main()


heartbeat.sh
#! /bin/sh

# heartbeat

### BEGIN INIT INFO
# Provides:          listen_for_shutdown.py
# Required-Start:    $remote_fs $syslog
# Required-Stop:     $remote_fs $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
### END INIT INFO

# If you want a command to always run, put it here

# Carry out specific functions when asked to by the system
case "$1" in
  start)
    echo "Starting heartbeat.py..."
    python3 /usr/local/bin/heartbeat.py &
    ;;
  stop)
    echo "Stopping heartbeat.py..."
    pkill -f /usr/local/bin/heartbeat.py
    ;;
    restart)
    echo "Restarting heartbeat.py..."
    stop
    start
    ;;
    status)
    echo "Status heartbeat.py..."
    PID=`pgrep -f /usr/local/bin/heartbeat.py`
    ps ${PID}
    ;;
  *)
    echo "Usage: /etc/init.d/heartbeat.sh {start|stop|status|restart}"
    exit 1
    ;;
esac

exit 0

juice4halt.py
# !/usr/bin/python3

"""
    juice4halt.py

    This is copied from listen_for_shutdown.py, which may be obsoleted in the future.
    The function 'button_press_callback' has been commented out along with references
    in the 'gpio_init' function.

    Flashes Yellow LED to indicate RPi is shutting down.

    https://howchoo.com/g/mwnlytk3zmm/how-to-add-a-power-button-to-your-raspberry-pi
    https://www.daniweb.com/programming/software-development/threads/507019/python-program-to-extract-ip-addresses-from-a-log-file
"""

__version__ = '1.0'
__author__ = 'Bill Doerr'
__date__ = '18Apr2021'


import RPi.GPIO as GPIO
import subprocess
import paramiko
import re
from time import sleep
import datetime
import logging
import threading


# Define some constants
CHANNEL_LED = 13            # Pinout #33, GPIO 13
TIMEOUT = 20                # Value in seconds
DELAY_SLOW = 0.25           # Value in seconds
DELAY_FAST = 0.1            # Value in seconds


# Log file initialization
def log_init():
    global logger

    logger = logging.getLogger(__name__)
    logger.setLevel(logging.INFO)

    formatter = logging.Formatter('%(asctime)s:%(levelname)s:\t%(message)s', '%H:%M:%S')

    file_handler = logging.FileHandler('/var/log/carputer/juice4halt.log', 'w')
    file_handler.setFormatter(formatter)

    logger.addHandler(file_handler)

    # Set up logging to console
    console = logging.StreamHandler()
    console.setLevel(logging.DEBUG)
    console.setFormatter(formatter)
    # logging.getLogger('').addHandler(console)
    logger.addHandler(console)

    logger.info("Logging initialized.")


# GPIO initialization
def gpio_init():
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(CHANNEL_LED, GPIO.OUT)

    logger.info("GPIO initialized.")


# Parses Ip's from dns lease file
def get_nodes():
    # filename = '/var/lib/misc/dnsmasq.leases'
    filename = '/etc/carputer/nodes.config'
    nodes = []
    try:
        leasefile = list(open(filename, 'r').read().split('\n'))
        logger.info("Lease file opened:  " + str(leasefile))
        for entry in leasefile:
            ips = re.findall(r'[0-9]+(?:\.[0-9]+){3}', entry)
            for ip in ips:
                logger.info("Ip added to nodes:  " + ip)
                nodes.append(ip)
    except IOError:
        logger.error("Unable to open file:  " + filename)
    finally:
        if nodes:   # empty sequences are false
            logger.info("Nodes returned:  " + str(nodes))
            return nodes
        else:
            logger.info("Nodes returned:  None")
            return None


# Sends shutdown command to specified ip
def shutdown_node(ip):
    ssh = paramiko.SSHClient()
    isconnected = False
    try:
        # Sure ain't very secure
        ssh.connect(ip, username='pi', password='scoobydoo', timeout=TIMEOUT)
        isconnected = True
    except:
        logger.info("SSH connection error.")

    if isconnected:
        try:
            # ssh_stdin, ssh_stdout, ssh_stderr = ssh.exec_command('sudo shutdown -h now', timeout=TIMEOUT)
            ssh.exec_command('sudo shutdown -h now', timeout=TIMEOUT)
        except:
            logger.info("SSH shutdown node error.")


# Loop through nodes and perform shutdown command
def shutdown_nodes():
    # Loop through nodes and send shutdown
    nodes = get_nodes()
    for node in nodes:
        logger.info("Shutting down node: " + str(node))
        shutdown_node(node)


# LED indicator that power off button pressed
def flash_led(e, mode):
    while not e.isSet():
        if mode == 1:   # Slow
            GPIO.output(CHANNEL_LED, GPIO.HIGH)
            sleep(DELAY_SLOW)
            GPIO.output(CHANNEL_LED, GPIO.LOW)
            sleep(DELAY_SLOW)
        elif mode == 2:     # Fast
            GPIO.output(CHANNEL_LED, GPIO.HIGH)
            sleep(DELAY_FAST)
            GPIO.output(CHANNEL_LED, GPIO.LOW)
            sleep(DELAY_FAST)

def led_flash():
    # Indicate we are shutting down
    e = threading.Event()
    logger.info("Indicate we are shutting down.")
    t = threading.Thread(name='non-block', target=flash_led, args=(e, 2))
    t.start()   # Start faster LED flashing

def main():

    log_init()
    gpio_init()
    
    # Indicate we are shutting down
    logger.info("Indicate we are shutting down.")
    led_flash()

    # Shutdown slave nodes
    logger.info("Shutting down slave nodes.")
    shutdown_nodes()
    # Shutdown master node
    logger.info("Shutting down master node.")
    subprocess.call(['sudo', 'shutdown', '-h', 'now'], shell=False)

    logger.info("Performing GPIO cleanup.")
    GPIO.output(CHANNEL_LED, GPIO.LOW)
    GPIO.cleanup()  # This ensures a clean exit.  Enables all pins for INPUT, but only for this program.


if __name__ == "__main__":
    main()


shutdown_script.sh
#!/bin/bash

#title           :shutdown_script.sh
#description     :This script runs on Raspberry Pi equipped with the Juice4Halt module.
#author		     :Nelectra s.r.o.
#date            :20190415
#version         :1.2 (bug fixed: collision with the rebootj4h script solved by reading the GPIO25 pin 2x) 
#usage		     :copy this file to /home/pi/juice4halt/bin/
#		          and make it executable: sudo chmod 755 shutdown_script 	
#notes           :www.juice4halt.com
#copyright	     :GNU GPL v3.0
#
#update_by      :Bill Doerr
#update_version :1.2.1
#update_date    :18Apr2021	
#==============================================================================

# create directory for working with GPIO and wait
echo "25" > /sys/class/gpio/export
sleep 0.2s
 
# use the pin as an open-drain output, send a short LOW pulse (= Booting finished, shutdown_script runs, automatic shutdown enabled)
# then change to input and listen
# if a LOW is sent from the J4H then start with the shutdown
 
# set GPIO25 as output and set output to LOW
echo "out" > /sys/class/gpio/gpio25/direction
echo "0" > /sys/class/gpio/gpio25/value
sleep 0.1s
echo "in" > /sys/class/gpio/gpio25/direction
 
# wait until pin rises to HI
sleep 0.1s
 
pinval1="1" # 1st reading
pinval2="1" # 2nd reading
 
# to be resistant to short pulses (e.g. 100ms pulse generated by the rebootj4h script) two readings 200ms apart from each other are required
 
while [ "$pinval1" !=  "0" -o "$pinval2" !=  "0" ]
do
        #reading the value of the input pin
        pinval1=$(cat /sys/class/gpio/gpio25/value)
       
        #wait
        sleep 0.2s
 
	    #reading the value of the input pin again 0.2s later
        pinval2=$(cat /sys/class/gpio/gpio25/value)

done
  
# set GPIO25 as output and set output to LOW
echo "out" > /sys/class/gpio/gpio25/direction
echo "0" > /sys/class/gpio/gpio25/value
 
# wait for system halt
# after system halted the pin will be automatically switched to input and the level will be pulled up to HI
# a LOW to HI transition signals to the J4H to turn the power off

echo ""
echo "Juice4halt: Power failure, RPi will now shut down."
# Changed by Bill Doerr. Make call to python script to shutdown slave nodes and master node
# sudo halt
python3 /usr/local/bin/juice4halt.py



listen_for_shutdown.sh
#! /bin/sh

#
# /etc/init.d
# listen_for_shutdown.sh
#

# https://werxltd.com/wp/2012/01/05/simple-init-d-script-template/
# https://howchoo.com/g/mwnlytk3zmm/how-to-add-a-power-button-to-your-raspberry-pi

### BEGIN INIT INFO
# Provides:          listen_for_shutdown.py
# Required-Start:    $remote_fs $syslog
# Required-Stop:     $remote_fs $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
### END INIT INFO

start() {
    echo "Starting listen_for_shutdown.py..."
    python3 /usr/local/bin/listen_for_shutdown.py &
    return
}

stop() {
    echo "Stopping listen_for_shutdown.py..."
    pkill -f /usr/local/bin/listen_for_shutdown.py
    return
}

status() {
    echo "Status listen_for_shutdown.py..."
    PID=`pgrep -f /usr/local/bin/listen_for_shutdown.py`
    ps ${PID}
    return
}

case "$1" in
  start)
        start
        ;;
  stop)
        stop
        ;;
  restart)
        stop
        start
    ;;
  status)
        status
    ;;
  *)
        echo "Usage: /etc/init.d/listen_for_shutdown.sh {start|stop|status|restart}"
        exit 1
    ;;
esac

exit 0


listen_for_shutdown.py
# !/usr/bin/python3

"""
    listen_for_shutdown.py

    Listens for button press on GPIO pin.

    After being pressed for ~5 seconds will send shutdown command
    to Ip's listed in the dnsmasq.leases file.  When this completes
    shutdown command will be sent to the node hosting this script.

    Flashes LED to indicate state of button press.

    https://howchoo.com/g/mwnlytk3zmm/how-to-add-a-power-button-to-your-raspberry-pi
    https://www.daniweb.com/programming/software-development/threads/507019/python-program-to-extract-ip-addresses-from-a-log-file
"""

__version__ = '1.1'
__author__ = 'Bill Doerr'


import RPi.GPIO as GPIO
import subprocess
import paramiko
import re
from time import sleep
import datetime
import logging
import threading


# Define some constants
CHANNEL_BUTTON = 26         # Pinout #37, GPIO #26
CHANNEL_LED = 13            # Pinout #33, GPIO 13
TIMEOUT = 20                # Value in seconds
BUTTON_PRESSED_TIME = 3     # Amount of time button must be pressed
DELAY_SLOW = 0.25           # Value in seconds
DELAY_FAST = 0.1            # Value in seconds


# Log file initialization
def log_init():
    global logger

    logger = logging.getLogger(__name__)
    logger.setLevel(logging.INFO)

    formatter = logging.Formatter('%(asctime)s:%(levelname)s:\t%(message)s', '%H:%M:%S')

    file_handler = logging.FileHandler('/var/log/carputer/listen_for_shutdown.log', 'w')
    file_handler.setFormatter(formatter)

    logger.addHandler(file_handler)

    # Set up logging to console
    console = logging.StreamHandler()
    console.setLevel(logging.DEBUG)
    console.setFormatter(formatter)
    # logging.getLogger('').addHandler(console)
    logger.addHandler(console)

    logger.info("Logging initialized.")


# GPIO initialization
def gpio_init():
    GPIO.setmode(GPIO.BCM)

    GPIO.setup(CHANNEL_BUTTON, GPIO.IN, pull_up_down=GPIO.PUD_UP)
    # GPIO.wait_for_edge(CHANNEL, GPIO.FALLING)
    GPIO.add_event_detect(CHANNEL_BUTTON, GPIO.FALLING, callback=button_press_callback, bouncetime=200)

    GPIO.setup(CHANNEL_LED, GPIO.OUT)

    logger.info("GPIO initialized.")


# Parses Ip's from dns lease file
def get_nodes():
    # filename = '/var/lib/misc/dnsmasq.leases'
    filename = '/etc/carputer/nodes.config'
    nodes = []
    try:
        leasefile = list(open(filename, 'r').read().split('\n'))
        logger.info("Lease file opened:  " + str(leasefile))
        for entry in leasefile:
            ips = re.findall(r'[0-9]+(?:\.[0-9]+){3}', entry)
            for ip in ips:
                logger.info("Ip added to nodes:  " + ip)
                nodes.append(ip)
    except IOError:
        logger.error("Unable to open file:  " + filename)
    finally:
        if nodes:   # empty sequences are false
            logger.info("Nodes returned:  " + str(nodes))
            return nodes
        else:
            logger.info("Nodes returned:  None")
            return None


# Sends shutdown command to specified ip
def shutdown_node(ip):
    ssh = paramiko.SSHClient()
    isconnected = False
    try:
        # Sure ain't very secure
        ssh.connect(ip, username='pi', password='scoobydoo', timeout=TIMEOUT)
        isconnected = True
    except:
        logger.info("SSH connection error.")

    if isconnected:
        try:
            # ssh_stdin, ssh_stdout, ssh_stderr = ssh.exec_command('sudo shutdown -h now', timeout=TIMEOUT)
            ssh.exec_command('sudo shutdown -h now', timeout=TIMEOUT)
        except:
            logger.info("SSH shutdown node error.")


# Loop through nodes and perform shutdown command
def shutdown_nodes():
    # Loop through nodes and send shutdown
    nodes = get_nodes()
    for node in nodes:
        logger.info("Shutting down node: " + str(node))
        shutdown_node(node)


# LED indicator that power off button pressed
def flash_led(e, mode):
    while not e.isSet():
        if mode == 1:   # Slow
            GPIO.output(CHANNEL_LED, GPIO.HIGH)
            sleep(DELAY_SLOW)
            GPIO.output(CHANNEL_LED, GPIO.LOW)
            sleep(DELAY_SLOW)
        elif mode == 2:     # Fast
            GPIO.output(CHANNEL_LED, GPIO.HIGH)
            sleep(DELAY_FAST)
            GPIO.output(CHANNEL_LED, GPIO.LOW)
            sleep(DELAY_FAST)


# Button press callback
def button_press_callback(channel):
    logger.info('Button pressed')

    # Removes event detection for a pin.
    GPIO.remove_event_detect(channel)

    # Flash LED
    # GPIO.output(CHANNEL_LED, GPIO.HIGH)
    e1 = threading.Event()
    e2 = threading.Event()
    t = threading.Thread(name='non-block', target=flash_led, args=(e1, 1))
    t.start()

    pressed_time_start = datetime.datetime.now()

    while not GPIO.input(channel):
        pressed_time_total = (datetime.datetime.now() - pressed_time_start).seconds

        logger.info("Button pressed time:  " + str(pressed_time_total))

        # Must hold button down for period of time
        # LED will flash faster once in shutdown mode
        if pressed_time_total > BUTTON_PRESSED_TIME:
            logger.info("Shutting down nodes.")
            e1.set()    # Stop slower LED flashing
            t = threading.Thread(name='non-block', target=flash_led, args=(e2, 2))
            t.start()   # Start faster LED flashing
            # Shutdown slave nodes
            shutdown_nodes()
            # Shutdown master node
            logger.info("Shutting down master node.")
            subprocess.call(['sudo', 'shutdown', '-h', 'now'], shell=False)
        else:
            logger.info("Button was not pressed long enough to perform shutdown.\n\t  "
                        "Button pressed time:  " + str(pressed_time_total))

    # Re-enable callback if button press too short or shutdown failed for some reason
    GPIO.add_event_detect(CHANNEL_BUTTON, GPIO.FALLING, callback=button_press_callback, bouncetime=200)
    GPIO.output(CHANNEL_LED, GPIO.LOW)

    # Disable LED flashing
    e1.set()
    e2.set()


def main():

    log_init()
    gpio_init()

    try:
        # Loop until shutdown
        logger.info("Starting main loop.")
        while True:
            sleep(1)

    except KeyboardInterrupt:
        logger.error("Keyboard interrupt.")

    except:
        # This catches ALL other exceptions including errors.
        # You won't get any error messages for debugging
        # so only use it once your code is working.
        logger.error("Unknown exception.")

    finally:
        logger.info("Performing GPIO cleanup.")
        GPIO.output(CHANNEL_LED, GPIO.LOW)
        GPIO.cleanup()  # This ensures a clean exit.  Enables all pins for INPUT, but only for this program.


if __name__ == "__main__":
    main()
